projects/python-client/src/datapane/client/api/report/core.py

```diff
@@ -465,3 +522,107 @@ def preview(
+        bundle_path = pl_path / SERVED_REPORT_BUNDLE_DIR
+        assets_path = pl_path / SERVED_REPORT_ASSETS_DIR
+
+        bundle_path.mkdir(parents=True, exist_ok=True)
+        assets_path.mkdir(parents=True, exist_ok=True)
+
+        self._served_local_writer.assert_bundle_exists()
+
+        # Copy across symlinked report bundle
+        copy_tree(str(self._served_local_writer.assets / "report"), str(bundle_path))
+
+        # Copy across symlinked Vue module
+        copyfile(str(self._served_local_writer.assets / VUE_ESM_FILE), str(pl_path / VUE_ESM_FILE))
+
+        local_doc, attachments = self._gen_report(embedded=False, served=True, title=name)
+
+        for a in attachments:
+            # TODO - compress in-memory to save a disk write?
+            # Copy across user report assets
+            with compress_file(a) as a_gz:
+                copyfile(str(a_gz), str(assets_path / Path(a).name))
+
+        self._served_local_writer.write(
+            local_doc,
+            str(pl_path / "index.html"),
+            name=name,
+            formatting=formatting,
+        )
+
+        display_msg(f"Successfully built report in {path}")
+
+    def serve(
+        self,
+        path: str,
+        port: int = 8000,
+        host: str = "localhost",
+        formatting: t.Optional[ReportFormatting] = None,
+        open: bool = False,
+    ):
+        """Serve the report from a local http server
+
+        Args:
+            path: File path to store the document; a new report will be built at the specified path if none is found
+            open: Open in your browser after creating (default: False)
+            port: The port used to serve the report (default: 8000)
+            host: The host used to serve the report (default: localhost)
+            formatting: Sets the basic report styling; note that this is ignored if a report exists at the specified path
+        """
+        if not osp.isdir(path):
+            self.build(path, formatting=formatting)
+
+        os.chdir(path)  # Run the server in the specified path
+        server = HTTPServer((host, port), DPServer)
+        display_msg(f"Server started at {host}:{port}")
+
+        if open:
+            # Polls the server in a new thread then opens on 200;
+            # if the endpoint is simply opened then there is a race
+            # between the page loading and server becoming available.
+            threading.Thread(target=self._open_server, args=(host, port), daemon=True).start()
+
+        try:
+            server.serve_forever()
+        except KeyboardInterrupt:
+            pass
+
+        server.server_close()
+
+    @staticmethod
+    def _open_server(host: str, port: int) -> None:
+        """Polls localserver endpoint until a 200 response is returned, then opens the report URL"""
+        url = f"http://{host}:{port}"
+        status_code: t.Optional[int] = None
+        request_limit = 100
+        attempts = 0
+
+        while status_code != 200:
```

Comment:
you can prob code-golf this a bit and remove several of the variables,

```python
for _ in range(1, 100):
  try:
    requests.get(url).raise_for_status()
  except requests.exceptions.RequestException:
    sleep(1)
    continue
else:
    raise TimeoutError(f"Reached maximum {request_limit} retries")
```

Reply:
Oh this is much nicer! Maybe not needed since I've simplified to daemon thread to just `sleep` for a while to yield to the main thread as described above

Reply:
Haha - it was fun to write even if not needed lol!